### PR TITLE
OneDimScene abstraction

### PR DESCRIPTION
**Pull request by leroyvn**
# Description

This PR introduces a new `OneDimScene` abstraction. It solves eradiate/eradiate-issues#13 and prepares the ground for solving eradiate/eradiate-issues#26.

Main change:

- Introducing `OneDimScene`: I added a `OneDimScene` class which is meant to replace the complicated scene setup code in the `OneDimSolverApp` post-init step. Thanks to attrs's fantastic init tools, I'm confident that we can remove Cerberus from the `OneDimSolverApp` configuration processing step! This class is unit-tested.

This was a large piece of work, which required the addition of a few components and gave me the opportunity to revisit parts of the code I had overlooked during past reviews.

Major changes:

- Introducing `Integrator`: for better control on integrator selection, I added a new `scenes.integrators` module. It supplies thin interfaces to Mitsuba's integrators with proper factory construction, making it possible to instantiate these components from a dictionary-based specification. This module is documented (with docstrings) and unit-tested.
- Introducing local `Measure` classes: I added local class definitions to specialise existing measures and make them more constrained. These measures, specifically designed to work with `OneDimSolverApp`, are not registered to the measure factory if the `onedim.app` module is not loaded. This will likely evolve over time, but it's a nice capability of our factory design. I did not add tests for these since they only provide minor interface changes and aliases to existing classes.

Minor changes:

- `Factory` update: `BaseFactory.register()` is now variadic and allows to register a class under several names (used here to alias measures).
- `Measure` fixes: I fixed outdated docstrings and class names in the `scene.measure` module. I also added a figure illustrating the plane radiancemeter array setup. Direction ordering and mapping to the result array is yet to be documented (also applies to the hemisphere radiancemeter array).
- `Atmosphere` fixes: I clarified and fixed confusing behaviour in the atmosphere height computation. Some unit castings were wrong and I clarified the meaning of "offset", "height" and "kernel height" (this is not documented completely but it will come).
- Bibliography: Added a reference and made a formatter pass.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is apropriately documented.
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

### CODE REVIEW

File: eradiate/solvers/onedim/app.py [hunk 1]
```diff
@@ -302,75 +496,79 @@ def _configure_scene(self):
...
-                        )
-
-                # Expand elements to kernel scene dictionary
-                self._kernel_dict.add(list(self._elements.values()))
+                    offset = atmosphere.kernel_offset.to(cdu.get("length")).magnitude
+                    height = atmosphere.kernel_height.to(cdu.get("length")).magnitude - offset
+
+                else:
+                    height = 0.1
+                    offset = 1e-3
```
Comment (nollety) [inline eradiate/solvers/onedim/app.py:530]: If `atmosphere` is `None`, I assume that `height` defines the top-of-atmosphere altitude where the sensor is placed for a `toa` measure, right? Why is`offset` not 0 though? There will not be any intersections issues without an atmosphere shape, if I understand correctly.

**Review by nollety — APPROVED** [review]
This system of "local measures" is really nice and makes a lot of sense!

File: eradiate/scenes/atmosphere/base.py [hunk 2]
```diff
@@ -82,7 +88,10 @@ def kernel_offset(self):
            This is required to ensure that the surface is the only shape
            which can be intersected at ground level during ray tracing.
         """
-        return self.kernel_height * 1e-3  # TODO: adjust offset based on medium profile
+        if self.height == "auto":
+            return ureg.Quantity(0.1, "km")  # This is a 0.1% offset for a height of 100 km
+        else:
+            return self.height * 1e-3  # TODO: adjust offset based on medium profile
```
[the remark this answers is not in the document]
Comment (leroyvn) [reply eradiate/scenes/atmosphere/base.py:94]: That's correct, there is an issue here. I'll think again about this: maybe I should add a member explicitly returning the TOA value. The problem here is that we need defaults to compute the kernel height and offset, but the offset is actually deduced from the height, and not the kernel height. To be clear: kernel height = height + kernel offset. I'll work on this further.

File: eradiate/scenes/atmosphere/heterogeneous.py [hunk 3]
```diff
@@ -229,9 +229,10 @@ def __attrs_post_init__(self):
     def kernel_height(self):
         if self.height == "auto":
             ds = self.profile.to_dataset()
-            return ureg.Quantity(ds.z_level.values.max(), ds.z_level.units)
+            return ureg.Quantity(ds.z_level.values.max(), ds.z_level.units) + \
+                   self.kernel_offset
         else:
-            return self.height
+            return self.height + self.kernel_offset
```
[the remark this answers is not in the document]
Comment (leroyvn) [reply eradiate/scenes/atmosphere/heterogeneous.py:235]: See above comment. There is indeed a tiny part of the medium underneath the surface (the offset is this famous kernel offset we've been talking about). I have a todo on my list to document that properly.

File: eradiate/scenes/integrators.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+        return result
+
+
+# Add doctrings
+MonteCarloIntegrator.__doc__ += f"""
+.. rubric:: Constructor arguments / instance attributes
+
+{_monte_carlo_integrator_params_docs}
+"""
```
[the remark this answers is not in the document]
Comment (leroyvn) [reply eradiate/scenes/integrators.py:102]: See issue eradiate/eradiate-issues#30: this is a rudimentary way of inheriting docstrings for data members (they don't have a `__doc__` member themselves so this must be worked around using the class's _`__doc__`.

**Review by schunkes — APPROVED** [review]

**Comment by leroyvn** [comment]
@yvan this PR is chained with #11, so I'll make a Schrödinger anwser: yes and no ;)

File: eradiate/solvers/onedim/app.py (hunk 1 above)
Comment (leroyvn) [reply eradiate/solvers/onedim/app.py:530]: That's a good question. I actually removed this use case later on, but maybe it should be reinstated? I'll think about that.

File: eradiate/scenes/atmosphere/heterogeneous.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (leroyvn) [reply eradiate/scenes/atmosphere/heterogeneous.py:235]: I added a small design note and a figure explaining what the different attributes are.

File: eradiate/scenes/atmosphere/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (leroyvn) [reply eradiate/scenes/atmosphere/base.py:94]: It's solved now: I changed the atmosphere's attributes and made `height` always return a numerical value. The attribute used to configure the object (and can be set to `"auto"`) is now `toa`.

**Comment by leroyvn** [comment]
Everything ready: merging!

**Pull request merged**